Source/WebCore/Modules/webaudio/BaseAudioContext.cpp

```diff
@@ -298,16 +299,18 @@ void BaseAudioContext::decodeAudioData(Ref<ArrayBuffer>&& audioData, RefPtr<Audi
     if (!m_audioDecoder)
         m_audioDecoder = makeUnique<AsyncAudioDecoder>();
 
-    m_audioDecoder->decodeAsync(WTFMove(audioData), sampleRate(), [this, activity = makePendingActivity(*this), successCallback = WTFMove(successCallback), errorCallback = WTFMove(errorCallback), promise = WTFMove(promise)](ExceptionOr<Ref<AudioBuffer>>&& result) mutable {
+    // We don't move audioData as we want to delete it on the main thread upon completion of the DecodingTaskPromise.
+    auto p = m_audioDecoder->decodeAsync(Ref { audioData }, sampleRate());
+    p->then(RunLoop::main(), __func__, [this, activity = makePendingActivity(*this), successCallback = WTFMove(successCallback), errorCallback = WTFMove(errorCallback), promise = WTFMove(promise), audioData = WTFMove(audioData)] (DecodingTaskPromise::Result&& result) mutable {
         queueTaskKeepingObjectAlive(*this, TaskSource::InternalAsyncTask, [successCallback = WTFMove(successCallback), errorCallback = WTFMove(errorCallback), promise = WTFMove(promise), result = WTFMove(result)]() mutable {
-            if (result.hasException()) {
+            if (!result) {
                 if (promise)
-                    promise.value()->reject(result.releaseException());
+                    promise.value()->reject(Exception { result.error(), "Decoding failed"_s });
                 if (errorCallback)
                     errorCallback->handleEvent(nullptr);
                 return;
             }
-            auto audioBuffer = result.releaseReturnValue();
+            auto audioBuffer = WTFMove(result.value());
             if (promise)
                 promise.value()->resolve<IDLInterface<AudioBuffer>>(audioBuffer.get());
```

Comment:
Could we use resolveWithNewlyCreated instead and move audioBuffer?

Reply:
That seems out of scope tbh.
The aim of this PR was to replace one existing code with an identical one using NativePromise.

Reply:
Actually, I don't think we can.
audioBuffer is returned to both the promise and `successCallback` you can't move it twice.